yaml/validate: Improve error reporting

kilpkonn opened the pull request:
This makes `yaml.Validate(data, schema)` work more similarly to what one would likely do manually (`schema.Unify(data).Validate()`).

Namely this improves error reporting.

Closes https://github.com/cue-lang/cue/issues/4002

kilpkonn committed:
yaml/validate: Add test for disjunction errors

Signed-off-by: Tavo Annus <tavo.annus@gmail.com>

mvdan commented:
Thanks! This change aligns with `pkg/encoding/json.Validate` so I think it's correct.

Please add a test; see `pkg/encoding/yaml/testdata/validate.txtar`.

Ideally we add the same sort of test to `pkg/encoding/json/testdata/gen.txtar` under the "validate" section, to sanity check that the two are consistent.

Typically, when adding a regression test, we add it before the fix to record the "before" output, via `CUE_UPDATE=1 go test ./...`, and then the second commit applies the fix and improves the expected output in the test. That way, it's very obvious how the fix is helping.

kilpkonn committed:
yaml/validate: Improve error reporting

This makes yaml.Validate work more similarly to what
one would likely do manually (`schema.Unify(data).Validate()`).

Namely this improves error reporting by including more suberrors.

Signed-off-by: Tavo Annus <tavo.annus@gmail.com>

kilpkonn commented:
Should be good now

mvdan reviewed (approved):
Thanks! Will import to Gerrit and merge.

mvdan commented:
Imported as https://review.gerrithub.io/c/cue-lang/cue/+/1219285; cc @myitcv for review
